src/node/connection_types.h

```diff
@@ -76,6 +77,9 @@ enum class ConnectionType {
     ADDR_FETCH,
 };
 
+/** Number of entries in ConnectionType. */
+static constexpr size_t NUM_CONNECTION_TYPES{6};
```

Comment:
Agree with the idea of programmatic calculation. I noticed [NET_MAX](https://github.com/bitcoin/bitcoin/blob/v25.0rc1/src/netaddress.h#L68) in `netaddress.h` accomplishes this by adding an enum value to the end of the list. That new value serves as a counter for all the enums. I initially implemented these (number of connection and message types) in that same manner because it seemed more programmatic. As long as new enums were inserted before, then I figured they would be accounted for in the total.

However, l Iater got feedback and the unreliability of mapping enums to integers was brought to my attention. Every single enum needs to be explicitly mapped to an integer, and if you're going to go through the trouble of that, then it comes out to be basically the same amount of work to make a standalone constant. There's also something that doesn't feel consistent about having an enum that accounts for the total living alongside the enum values that are being counted.

In this PR I have a `static assert` checking that `NUM_NET_MESSAGE_TYPES` is what we expect it to be. Would love if there was somehow a way to do that with the number of connection types as well. It's not exactly setting the value in a programmatic way, but it's at least providing some kind of check.

From `src/protocol.cpp`:
```
static_assert(NUM_NET_MESSAGE_TYPES == sizeof(allNetMessageTypes) / sizeof(allNetMessageTypes[0]), "Please update NUM_NET_MESSAGE_TYPES");
```

Reply:
@amitiuttarwar sort of. It's two things:

1. If enum values are not explicitly assigned (which is the case for the `Network` enum right now), there is no guarantee that `NET_MAX` is what you think it will be. This came out of a conversation with @vasild during review earlier this year. It's my understanding that `NET_MAX` _probably_ corresponds to the total number of enum values, but even today there are no guarantees on the value of `NET_MAX`. I'd like to do a little more research here and get some clear documentation on this though, as I'm currently just going off of information that was given to me.

2. If you start explicitly assigning integers to each enum value, you still have to update `NET_MAX`, so I don't think it is in that much danger of falling out of sync, but it is the same amount of work as updating a constant like `NUM_CONNECTION_TYPES` here.